pkg/services/hcloud/server/server.go

```diff
@@ -190,6 +217,198 @@ func (s *Service) handleBootStateUnset(ctx context.Context) (reconcile.Result, e
 	return reconcile.Result{RequeueAfter: requeueAfter}, nil
 }
 
+func (s *Service) handleBootStateWaitForPreRescueOSThenEnableRescueSystem(ctx context.Context, server *hcloud.Server) (res reconcile.Result, reterr error) {
+	// analyze status of server
+	switch server.Status {
+	case hcloud.ServerStatusStarting, hcloud.ServerStatusInitializing:
+		conditions.MarkFalse(
+			s.scope.HCloudMachine,
+			infrav1.ServerAvailableCondition,
+			infrav1.ServerStartingReason,
+			clusterv1.ConditionSeverityInfo,
+			"server is %s", server.Status,
+		)
+		return reconcile.Result{RequeueAfter: requeueIntervalWaitForPreRescueOSThenEnableRescueSystem}, nil
+	case hcloud.ServerStatusRunning:
+		// execute below code
+	default:
+		// some temporary status
+		ctrl.LoggerFrom(ctx).Info("Unknown hcloud server status", "status", server.Status)
+		return reconcile.Result{RequeueAfter: requeueIntervalWaitForPreRescueOSThenEnableRescueSystem}, nil
+	}
+
+	// Server is Running.
+
+	_, hcloudSSHKeys, err := s.getSSHKeys(ctx)
+	if err != nil {
+		return res, fmt.Errorf("getSSHKeys failed: %w", err)
+	}
+
+	rescueOpts := &hcloud.ServerEnableRescueOpts{
+		Type:    hcloud.ServerRescueTypeLinux64,
+		SSHKeys: hcloudSSHKeys,
+	}
+	result, err := s.scope.HCloudClient.EnableRescueSystem(ctx, server, rescueOpts)
```

Comment:
why do we reboot in "enable rescue system" when we have a state "wait for rescue enabled then reboot"? 

Seems as if the states are misleading.. The current state says it will just enable rescue, but it also reboots..

I especially don't understand this really because in the next state all we do is reboot again

Reply:
@guettli I still believe this is the case...

Reply:
I'm not sure if you are aware that you trigger a reboot already in EnableRescueSystem